newsrevenuehub/rev-engine
# Dev 1420 db review apps

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

This is draft, I guess, in that not all of the requirements are met yet. But I may want to defer some of those because I'm not sure I know how to meet them, without help.

#### What's this PR do?

For each open PR / Review App:
- makes contribution pages automatically available at a ticket-specific URL. e.g https://civilbeat-dev-1420.revengine-review.org/donate 
- RP slugs are modified to include the ticket 
- Stripe webhook is created
- DNS entries are created
- Related Config Vars are created/updated
- This is all idempotent such that if the commands are run multiple times dupes aren't created
- database is populated (from a configurable source)

- A teardown script is there to remove the parts that we'd like to remove after the PR is closed (DNS, Webhooks)

#### Why are we doing this? How does it help us?

- to make Review Apps more immediately useful after creation
- to make it so others can log in and test features/bug fixes before they're merged 
- to save developers a lot of time bootstrapping things in order to test

#### How should this be manually tested?

"lite" version:
1. Visit a contribution page like: https://civilbeat-dev-1420.revengine-review.org/donate
2. Confirm you can make a contribution
3. Confirm you can log on to the Django admin at https://dev-1420.revengine-review.org/nrhadmin/ with your credentials from test
4. Confirm that the contribution you made show there; approve if necessary
5. Confirm that the Stripe details are there (confirming the webhook is working)
6. Confirm you can reset your password there via https://dev-1420.revengine-review.org/nrhadmin/password_reset/

"full" version:
1. create another branch from this one (`DEV-1420_db_review_apps`). Name it something that looks like a ticket: e.g `ABC-1234_foo_bar_baz`
2. Open a PR. 
3. Wait for Heroku to build a Review App an run the post-deployment scripts. 
4. Test the items in the lite version but using the fake ticket number: e.g. https://civilbeat-abc-1234.revengine-review.org/donate
5. Use `heroku run` to run the script directly: `python manage.py bootstrap-review-app`
6. Also run the tear down script `python manage.py teardown-review-app`

#### Have automated unit tests been added?

Yes.

#### How should this change be communicated to end users?

Let the product/engineering teams know.

#### Are there any smells or added technical debt to note?

- I wasn't able to get the Org Admin working because of the "can't find slug" thing. I attempted to fix that on line 12 of `spa/src/settings.js` but it's not working. If someone can assist here I'd appreciate it.
- There's a Heroku hook to run a tear down script when a PR is closed and Review App is deleted, but it's not supported for the container stack, which is unfortunate. So it'll need to be run manually. It can be run elsewhere just supplying the ticket ID so I suspect we'll do that as a manual step after merging a PR or periodically.

#### Has this been documented? If so, where?

I'll document in the wiki after merge.

#### What are the relevant tickets?

[DEV-1420]

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

Some changes have already been made to the pipeline variables to support this, but I none of them apply to production where none of this will run so I don't think anything needs to change there.

#### Are there next steps? If so, what? Have tickets been opened for them?

- https://news-revenue-hub.atlassian.net/browse/DEV-1541

#### Are there any security implications for this ticket?


[DEV-1420]: https://news-revenue-hub.atlassian.net/browse/DEV-1420?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
